rabbitmq/consumer.go

```diff
@@ -26,51 +26,85 @@ import (
 	"github.com/streadway/amqp"
 )
 
-// A consumer that is works with Handler interface
-// It needs a RabbitMQClient to work with and is started with the Run() method
-type RabbitMQConsumer struct {
+type ConsumerConfig struct {
+	// PrefetchCount configures how many in-flight "deliveries" are available to the consumer to ack/nack.
+	// ref: https://www.rabbitmq.com/consumer-prefetch.html
+	// There's no default value for the reason that it's very easy to misuse RMQ and have multiple consumers
+	// with too many deliveries in flight which results into badly distributed work load and high memory footprint
+	// of the consumers.
+	PrefetchCount int
+}
+
+type Consumer struct {
 	client  *RabbitMQClient
-	done    chan bool
 	handler Handler
 	logger  logger.StructuredLogger
 	metric  Metric
+	cfg     ConsumerConfig
 }
 
 func NewConsumer(
 	client *RabbitMQClient,
 	handler Handler,
 	logger logger.StructuredLogger,
 	metric Metric,
-) *RabbitMQConsumer {
-	return &RabbitMQConsumer{
+	cfg ConsumerConfig,
+) *Consumer {
+	return &Consumer{
 		client:  client,
-		done:    make(chan bool),
 		handler: handler,
 		logger:  logger,
 		metric:  metric,
+		cfg:     cfg,
 	}
 }
 
-func (c *RabbitMQConsumer) Run(ctx context.Context) error {
+func (c *Consumer) Run(ctx context.Context) error {
+	channel, err := c.client.CreateChannel(ctx)
+	if err != nil {
+		return stacktrace.Propagate(err, "failed to create a RMQ channel")
+	}
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	closeCh := channel.NotifyClose(make(chan *amqp.Error))
+
 	go func() {
-		<-ctx.Done()
-		c.logger.Info("Received context cancel. Going to close rabbit connections.")
-		_ = c.client.channel.Cancel(c.handler.GetConsumerTag(), false)
+		select {
+		case rmqErr := <-closeCh:
+			cancelFunc()
+			c.logger.Warn(
+				"RMQ closed the connection",
+				zap.String("reason", rmqErr.Reason),
+				zap.Int("code", rmqErr.Code),
+				zap.Bool("recover", rmqErr.Recover),
+				zap.Bool("server", rmqErr.Server),
+			)
+			return
+		case <-ctx.Done():
+			c.logger.Info("Received context cancel. Going to close RMQ connections.")
+			_ = channel.Cancel(c.handler.GetConsumerTag(), false)
+
+			if !c.handler.WaitToConsumeInflight() {
+				_ = channel.Close()
+			}
 
-		if !c.handler.WaitToConsumeInflight() {
-			c.client.channel.Close()
+			c.logger.Info("RMQ consumer stopped.")
+			_ = c.client.Close()
 		}
-
-		<-c.done
-		c.logger.Info("handler stopped")
-		_ = c.client.Close()
 	}()
 
 	if ctx.Err() != nil {
```

Comment:
I may be missing something here but wouldn't this be nil (almost) always as the goroutine which we start above might get an error on the context a bit later whereby the execution of the parent (current) function would usually have `ctx.Err() == nil`. That is, shouldn't we get the `ctx.Err()` when we receive a value on the `<-ctx.Done()` channel?

Reply:
> may be missing something here but wouldn't this be nil (almost) always as the goroutine which we start above might get an error on the context a bit later 

The context still may be canceled at this point of time. 
While the chance of this happening is very slim, the general rule-of-thumb is to check for context cancellation before committing to doing expensive I/O work, such as the communication with RMQ.

On the other hand checking before every I/O call is a bit too much, hence why we do it at this point.